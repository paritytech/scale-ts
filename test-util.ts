/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
