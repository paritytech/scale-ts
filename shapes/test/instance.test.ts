import * as $ from "../../mod.ts"
import { testInvalid, testShape } from "../../test-util.ts"

// Doesn't extend `Error` to avoid call-stack instability
class MyError {
  constructor(
    readonly code: number,
    readonly message: string,
    readonly payload: {
      a: string
      b: number
      c: boolean
    },
  ) {}
}

const $myError = $.withMetadata(
  $.metadata("$myError"),
  $.instance(
    MyError,
    $.tuple(
      $.u8,
      $.str,
      $.object(
        $.field("a", $.str),
        $.field("b", $.u8),
        $.field("c", $.bool),
      ),
    ),
    (myError: MyError) => [myError.code, myError.message, myError.payload],
  ),
)

testShape($myError, [
  new MyError(
    1,
    "At war with my Arch system config",
    {
      a: "a",
      b: 2,
      c: true,
    },
  ),
])

testInvalid($myError, [
  null,
  undefined,
  { code: 123, message: "foo", payload: { a: "abc", b: 2, c: true } },
  Object.assign(new Error("foo"), { stack: "Error: foo" }),
  new MyError(-1, "a", { a: "abc", b: 2, c: true }),
  new MyError(123, "a", { a: "abc", b: 2, c: "idk" } as any),
])
