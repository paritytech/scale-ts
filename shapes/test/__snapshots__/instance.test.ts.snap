export const snapshot = {};

snapshot[`\$myError MyError {
  code: 1,
  message: "At war with my Arch system config",
  payload: { a: "a", b: 2, c: true } } 1`] = `
01
84
41
74
20
77
61
72
20
77
69
74
68
20
6d
79
20
41
72
63
68
20
73
79
73
74
65
6d
20
63
6f
6e
66
69
67
04
61
02
01
`;

snapshot[`\$myError invalid null 1`] = `ShapeAssertError: !(value instanceof MyError)`;

snapshot[`\$myError invalid undefined 1`] = `ShapeAssertError: !(value instanceof MyError)`;

snapshot[`\$myError invalid { code: 123, message: "foo", payload: { a: "abc", b: 2, c: true } } 1`] = `ShapeAssertError: !(value instanceof MyError)`;

snapshot[`\$myError invalid [Error: foo] 1`] = `ShapeAssertError: !(value instanceof MyError)`;

snapshot[`\$myError invalid MyError { code: -1, message: "a", payload: { a: "abc", b: 2, c: true } } 1`] = `ShapeAssertError: value#arguments[0] < 0`;

snapshot[`\$myError invalid MyError { code: 123, message: "a", payload: { a: "abc", b: 2, c: "idk" } } 1`] = `ShapeAssertError: typeof value#arguments[2].c !== "boolean"`;
